jerry-core/ecma/builtin-objects/ecma-builtin-global.c

```diff
@@ -474,79 +471,25 @@ ecma_builtin_global_object_encode_uri_helper (lit_utf8_byte_t *input_start_p, /*
     {
       if (ecma_builtin_global_object_character_is_in (octets[0], unescaped_uri_bitset_p))
       {
-        output_length++;
+        ecma_stringbuilder_append_byte (&builder, octets[0]);
       }
       else
       {
-        output_length += URI_ENCODED_BYTE_SIZE;
-      }
-    }
-    else
-    {
-      output_length += utf_size * URI_ENCODED_BYTE_SIZE;
-    }
-  }
-
-  ecma_value_t ret_value;
-
-  JMEM_DEFINE_LOCAL_ARRAY (output_start_p,
-                           output_length,
-                           lit_utf8_byte_t);
-
-  lit_utf8_byte_t *output_char_p = output_start_p;
-  input_char_p = input_start_p;
-
-  while (input_char_p < input_end_p)
-  {
-    /* Input decode. */
-    input_char_p += lit_read_code_unit_from_utf8 (input_char_p, &ch);
-    cp = ch;
-
-    if (lit_is_code_point_utf16_high_surrogate (ch))
-    {
-      ecma_char_t next_ch;
-      lit_utf8_size_t read_size = lit_read_code_unit_from_utf8 (input_char_p, &next_ch);
-
-      if (lit_is_code_point_utf16_low_surrogate (next_ch))
-      {
-        cp = lit_convert_surrogate_pair_to_code_point (ch, next_ch);
-        input_char_p += read_size;
-      }
-    }
-
-    lit_utf8_size_t utf_size = lit_code_point_to_utf8 (cp, octets);
-
-    if (utf_size == 1)
-    {
-      if (ecma_builtin_global_object_character_is_in (octets[0], unescaped_uri_bitset_p))
-      {
-        *output_char_p++ = octets[0];
-      }
-      else
-      {
-        ecma_builtin_global_object_byte_to_hex (output_char_p, octets[0]);
-        output_char_p += URI_ENCODED_BYTE_SIZE;
+        ecma_builtin_global_object_byte_to_hex (result_chars, octets[0]);
+        ecma_stringbuilder_append_raw (&builder, result_chars, 3);
       }
     }
     else
     {
       for (uint32_t i = 0; i < utf_size; i++)
       {
-        ecma_builtin_global_object_byte_to_hex (output_char_p, octets[i]);
-        output_char_p += URI_ENCODED_BYTE_SIZE;
+        ecma_builtin_global_object_byte_to_hex (result_chars, octets[i]);
```

Comment:
Move the declaration of `result_chars` here (inside the for loop). Also assert that the `utf_sizte` is not greater that the size of the buffer.